src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -515,6 +502,54 @@ async def build_pex(
     )
 
 
+def _validate_metadata(
+    metadata: LockfileMetadata, request: PexRequest, python_setup: PythonSetup
+) -> None:
+    validation = metadata.is_valid_for(
+        request.requirements.lockfile_hex_digest,
+        request.interpreter_constraints,
+        python_setup.interpreter_universe,
+    )
+
+    if validation:
+        return
+
+    message_parts = [
+        f"Invalid lockfile for PEX request `{request.output_filename}`.",
+        "\n",
+    ]
+
+    if InvalidLockfileReason.INVALIDATION_DIGEST_MISMATCH in validation.failure_reasons:
+        message_parts.append(
+            "The requirements set for this PEX request are different to the requirements set when "
+            "the lockfile was generated. To fix this, you will need to regenerate the lockfile. "
+        )
+        message_parts.append(
+            f"(Expected requirements digest: {request.requirements.lockfile_hex_digest}, "
+            "actual: {metadata.requirements_invalidation_digest})"
+        )
+        message_parts.append("\n")
+
+    if InvalidLockfileReason.INTERPRETER_CONSTRAINTS_MISMATCH in validation.failure_reasons:
+        message_parts.append(
+            "The lockfile was generated under different interpreter constraints to the constraints "
+            "that are set. If you have overridden your project's interpreter constraints, you can "
```

Comment:
It'd be great to put in parentheses what the ICs are in the lockfile vs. the context.

Also, the issue really is ICs not being _compatible_, rather than being "different". Maybe something like:

> The requested interpreter constraints (<blah>) are not compatible with the lockfile's interpreter constraints (<blah>). Please either update the interpreter constraints to be compatible with the lockfile's, or regenerate the lockfile.